maaslin2 version update to 1.16.0

renu-pal opened the pull request:
FOR CONTRIBUTOR:
[* ] I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [ ] License permits unrestricted use (educational + commercial)
* [ ] This PR adds a new tool or tool collection
[* ] This PR updates an existing tool or tool collection
* [ ] This PR does something else (explain below)

renu-pal commented:
I am getting the following  error while passing multiple reference factors. 
![Screenshot from 2024-05-07 20-06-37](https://github.com/galaxyproject/tools-iuc/assets/148717065/d10aabaf-06a9-4ef6-ba4d-2521c008d95a)

Also any idea why residuals output is empty ?

paulzierep commented on tools/maaslin2/maaslin2.xml:
```diff
@@ -338,6 +345,7 @@ cd outputFolder && mkdir -p figures/ && cp *.pdf figures
             <param name="input_data" value="HMP2_taxonomy.tsv"/>
             <param name="input_metadata" value="HMP2_metadata.tsv"/>
             <param name="random_effects" value="3" />
+            <param name="reference" value="site,MGH;diagnosis,nonIBD"/>
```
Could the `;` be the issue with the reference ? Can you check: https://github.com/biobakery/biobakery/wiki/maaslin2 
The reference needs to be part of the fixed_effects.

renu-pal replied:
![Screenshot from 2024-05-10 11-37-20](https://github.com/galaxyproject/tools-iuc/assets/148717065/d167a948-f2f2-4a5c-abd1-b2cd6673b7a9)

They have mentioned that semicolon should be used for multiple factors, but have not given any example for it anywhere. Also if fixed column is null then the default value , i,e "all values" is chosen, so diagnosis and site are part of fixed_effects in this test case.
Is there a possibility that more than one reference factor is never used in practice?

renu-pal commented:
I have removed the test case 4 and 5 with fixed_effect=NULL. Please let me know if I am missing something else.

bernt-matthias commented on tools/maaslin2/maaslin2.xml:
```diff
@@ -64,6 +64,9 @@ Maaslin2.R
     --fixed_effects \$fixed_effects
 #end if
+#if $reference
+    --reference $reference
```
The parameter value should be single quoted.

renu-pal replied:
Thanks for pointing out.

bernt-matthias commented on tools/maaslin2/maaslin2.xml:
```diff
@@ -173,13 +177,13 @@ cd outputFolder && mkdir -p figures/ && cp *.pdf figures
             <output name="residuals">
                 <assert_contents>
-                    <has_size value="462746" delta="1000"/>
+                    <has_size value="0" />
```
Why is this becoming empty?

renu-pal replied:
I had the same doubt. It was not the case in the previous version.

bernt-matthias commented on tools/maaslin2/maaslin2.xml:
```diff
@@ -334,100 +341,9 @@ cd outputFolder && mkdir -p figures/ && cp *.pdf figures
                 </element>                                                                    
             </output_collection>
         </test>   
-        <test expect_num_outputs="5">
```
Why remove a test?

renu-pal replied:
I was not able to pass multiple reference(for site and diagnosis  both ) using semi colon as delimiter , as has been mentioned in the tutorials. I asked the same doubt in the forum. So removed the test cases till they do not resolve the query. 

https://forum.biobakery.org/t/passing-multiple-reference-factors-in-maaslin2-version-1-16-0/7035

renu-pal commented on tools/maaslin2/maaslin2.xml:
```diff
@@ -85,11 +88,8 @@ cd outputFolder && mkdir -p figures/ && cp *.pdf figures
-        	
-       
-        	
+        <param argument="--reference" type="text"  label="Reference" help="Reference for a variable with more than two levels provided as a string of ’variable,reference’ semi-colon delimited for multiple variables. " />
```
setting it as optional does not resolve the issue. According to resources the default value of reference is NA, but it gives error when I don't pass any value in the reference. 

![Screenshot from 2024-05-26 22-01-11](https://github.com/galaxyproject/tools-iuc/assets/148717065/df374b94-b908-41d4-802f-e85237e25e39)

https://github.com/biobakery/Maaslin2

I also posted a query about the same

https://forum.biobakery.org/t/passing-multiple-reference-factors-in-maaslin2-version-1-16-0/7035

renu-pal commented:
Hi @paulzierep , @bernt-matthias 
I have made some changes in the test cases , as setting fixed_effect =NULL was taking infinite time to run, so I changed it. 
Can you please go through this and let me know if its ready to go now?

bernt-matthias reviewed (commented):
Appreciate your efforts. But I'm still not happy with the output getting empty in the tests. My guess is that `outputFolder/residuals.rds` is not produced (maybe the name changed)? Also the justification of removing the last test is not clear to me.

FYI https://github.com/bioconda/bioconda-recipes/pull/48229

renu-pal commented:
Hi @bernt-matthias 
Thanks for raising doubts. 
Yes the request was answered and I have depicted the use of multiple characteristic variable in 4th test case. Turned out, comma was to be used instead of semi-colon. 
Also in the previous version, we added last two test cases where the  value of fixed_effects and Random_effects was set  NULL in test case 4 and just fixed_effects was NULL in 5th test case. 
In this version, setting fixed_effects as NULL does not work somehow. The test command takes infinite time to run. So I removed those test cases and added a new test case showing use of reference. 

About the residuals being empty, I will raise an issue again.

renu-pal commented:
Hi @paulzierep , @bernt-matthias  
I have updated the test cases accordingly but I am unable to understand why 2 checks are failing . Can you please guide me on this?

bernt-matthias commented:
Independent of this: can you switch to the `bioconductor-maaslin2` requirement: https://github.com/bioconda/bioconda-recipes/pull/48229

bernt-matthias commented:
> I have updated the test cases accordingly but I am unable to understand why 2 checks are failing 

Me neither. R errors messages .. as usual completely unusable.

paulzierep commented:
It seems that this is the error we get for all tests:
```
Error in checkForRemoteErrors(val) : 
  4 nodes produced errors; first error: object 'current_args' not found
Calls: Maaslin2 ... clusterApply -> staticClusterApply -> checkForRemoteErrors
Execution halted
```
Ones again this is something we should check in the biobakery forum. 

You said that only 2 checks are failing, could you show me an example where the test works? Does the tool work locally ? This might actually be a bioconda issue ? Can you try with  `bioconductor-maaslin2` as suggested ?

paulzierep commented on tools/maaslin2/macros.xml:
```diff
@@ -18,6 +18,7 @@
     <xml name="requirements">
         <requirements>
             <requirement type="package" version="@TOOL_VERSION@">maaslin2</requirement>
```
```suggestion
            <requirement type="package" version="@TOOL_VERSION@">bioconductor-maaslin2</requirement>
```

paulzierep replied:
@renu-pal can you commit this and check if the error persists ? Locally this worked.

paulzierep reviewed (commented):
Change conda package

paulzierep commented:
@renu-pal Can you change `\${GALAXY_SLOTS:-4}` to 1 and push before our meeting please ?

paulzierep commented:
I did a rather stupid test locally, but using `--cores 1` works and using `--cores 4` fails with the given error. It seems that there is something internally with the R code that cannot work with multithreading. 
Interestingly, the GitHub CI seems to assign 4 to `\${GALAXY_SLOTS:-4}` and locally planemo gives 1 to `\${GALAXY_SLOTS:-4}`; not sure how this can be influenced locally. But I think for now we could quickly fix the tool by using only 1 thread. And check with the biobakery folks if they have a fix for this.

paulzierep commented:
Lets see: https://forum.biobakery.org/t/maaslin2-1-16-0-fails-with-more-then-1-cpu/7172

bernt-matthias commented:
> not sure how this can be influenced locally

You can just call `GALAXY_SLOTS=4 planemo test ...`

renu-pal commented:
Hi @paulzierep , @bernt-matthias 
Finally all the checks have passed! Can this be merged now so I can start working on the tutorial :)

paulzierep commented:
@bernt-matthias @bgruening wdyt good to merge ? The limit to 1 cpu is not ideal, but we can update when the devs fix it ...

bernt-matthias commented on tools/maaslin2/maaslin2.xml:
```diff
@@ -73,23 +77,24 @@ Maaslin2.R
 #end if
     $output.plot_scatter
-    --cores \${GALAXY_SLOTS:-4}
+    --cores 1
```
```suggestion
    ## see https://forum.biobakery.org/t/maaslin2-1-16-0-fails-with-more-then-1-cpu/7172
    --cores 1
```

bernt-matthias reviewed (approved):
Feel free to add the suggested change if you like. Then I can merge tomorrow.

Note that if there are Galaxy servers that enabled more resources (TPV) they will now waste them.

paulzierep commented:
> Feel free to add the suggested change if you like. Then I can merge tomorrow.
> 
> Note that if there are Galaxy servers that enabled more resources (TPV) they will now waste them.

You mean if they assign more CPUs to maaslin2 via TPV ? we could add an entry here: https://github.com/galaxyproject/tpv-shared-database/blob/main/tools.yml I guess, but should not forget to remove it, ones it's fixed ...
